core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -620,6 +631,23 @@ private Optional<Integer> maybeLeaderEpoch(int leaderEpoch) {
         return leaderEpoch == RecordBatch.NO_PARTITION_LEADER_EPOCH ? Optional.empty() : Optional.of(leaderEpoch);
     }
 
+    public AsyncOffsetReadFutureHolder<Either<Exception, Option<FileRecords.TimestampAndOffset>>> asyncOffsetRead(
+            TopicPartition topicPartition,
+            Long timestamp,
+            Long startingOffset,
+            LeaderEpochFileCache leaderEpochCache,
+            Supplier<Option<FileRecords.TimestampAndOffset>> searchLocalLog) {
+        CompletableFuture<Either<Exception, Option<FileRecords.TimestampAndOffset>>> taskFuture = new CompletableFuture<>();
+        Future<Void> jobFuture = remoteStorageReaderThreadPool.submit(
+                new RemoteLogOffsetReader(this, topicPartition, timestamp, startingOffset, leaderEpochCache, searchLocalLog, result -> {
+                    TopicPartitionOperationKey key = new TopicPartitionOperationKey(topicPartition.topic(), topicPartition.partition());
+                    taskFuture.complete(result);
+                    delayedRemoteListOffsetsPurgatory.checkAndComplete(key);
+                })
+        );
+        return new AsyncOffsetReadFutureHolder<>(jobFuture, taskFuture);
```

Comment:
Thanks for the review! 

The reason for maintaining 2 futures: jobFuture and taskFuture. They are required to trigger the delayed operation completion (delayedRemoteListOffsetsPurgatory#checkAndComplete(key)) in the same remote-log-reader thread. When all the partitions results are received, then the delayedOperation gets completed.

Reply:
Thanks for the explanation, but "trigger the delayed operation completion in the same thread" seems to work by `CompletableFuture.supplyAsync`, right?
```java
        CompletableFuture<Either<Exception, Option<FileRecords.TimestampAndOffset>>> taskFuture = CompletableFuture.supplyAsync(() -> {
            Either<Exception, Option<FileRecords.TimestampAndOffset>> result;
            try {
                // If it is not found in remote storage, then search in the local storage starting with local log start offset.
                Option<FileRecords.TimestampAndOffset> timestampAndOffsetOpt =
                        OptionConverters.toScala(findOffsetByTimestamp(topicPartition, timestamp, startingOffset, leaderEpochCache))
                                .orElse(searchLocalLog::get);
                result = Right.apply(timestampAndOffsetOpt);
            } catch (Exception e) {
                // NOTE: All the exceptions from the secondary storage are catched instead of only the KafkaException.
                LOGGER.error("Error occurred while reading the remote log offset for {}", topicPartition, e);
                result = Left.apply(e);
            } finally {
                TopicPartitionOperationKey key = new TopicPartitionOperationKey(topicPartition.topic(), topicPartition.partition());
                delayedRemoteListOffsetsPurgatory.checkAndComplete(key);
            }
            return result;
        }, remoteStorageReaderThreadPool);
```

I notice there is another similar pattern `DelayedRemoteFetch`, so it is OK to keep current design for consistency. However, it would be great to let me known (for my own education) what the side-effect happens if using `CompletableFuture.supplyAsync` :)

Reply:
This won't work. When `delayedRemoteListOffsetsPurgatory.checkAndComplete(key)` is invoked, it calls `DelayedRemoteListOffset#tryComplete` which checks whether the taskFuture gets completed or not. It is not completed so the delayedOperation won't complete.

Reply:
> It is not completed so the delayedOperation won't complete.

oh, you are totally right, thanks!!!